File: leetcode/two_sum.rb
require 'pry'
# iterate thru array
# take each number and test adding it against the remaining numbers in the array
# keep track of indexes
# test the sum against the target
# if it true, stop and spit back the indicies
# def two_sum(nums, target)
#   nums.each_with_index do |value, index|
#     nums[(index+1)..-1].each do |num|
#       if value + num == target
#         return [index, nums.index(num)]
#       end
#     end
#   end
# end
# # nums = [2,7,11,15]
# # target = 9
# nums = [3,3]
# target = 6
# answr =  two_sum(nums, target)
# binding.pry

def two_sum(nums, target)
  pairs = nums.map.with_index {|value,index| [value,index]}
  pairs.each_with_index do |pair, index|
    pairs[(index + 1)..-1].each do |pair2|
      if pair[0] + pair2[0] == target
        return [pair[1], pair2[1]]
      end
    end
  end
end
# nums = [2,7,11,15]
# target = 9
nums = [3,3]
target = 6
answr =  two_sum(nums, target)
binding.pry

# cleaner solution from internets
# def two_sum(nums, target)
#   hash = {}
#   nums.each_with_index do |number, index|
#       if hash[target - number]
#           return [hash[target - number], index]
#       else
#           hash[number] = index
#       end
#   end
# end


# Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

# You may assume that each input would have exactly one solution, and you may not use the same element twice.

# You can return the answer in any order.
 
# Example 1:
# Input: nums = [2,7,11,15], target = 9
# Output: [0,1]
# Output: Because nums[0] + nums[1] == 9, we return [0, 1].

# Example 2:
# Input: nums = [3,2,4], target = 6
# Output: [1,2]

# Example 3:
# Input: nums = [3,3], target = 6
# Output: [0,1]

# Constraints:
#     2 <= nums.length <= 103
#     -109 <= nums[i] <= 109
#     -109 <= target <= 109
#     Only one valid answer exists.

File: leetcode/two_sum.js
// Given an array of integers nums and an integer target, 
// return indices of the two numbers such that they add up to target.

// nums = [2,7,11,15], target = 9

// for each item in the Array
// remember its index
// subtract it from the target
// scan the rest of the array for that value
// and if the value is found, return its array along with the item that worked

// var twoSum = function(nums, target) {
//   let final = [];

//   for (let index = 0; index < nums.length; index++) {
//     const element = nums[index];
//     const difference = target - element;
    
//     if (nums.slice(index+1).findIndex(e => e === difference) !== -1) {
//       final = [index, nums.findIndex(e => e === difference)];
//       return final;
//     }
//   }
// };

// let nums = [2,7,11,15]
// let target = 9
// let nums = [3,2,4]
// let nums = [3,3]
// let target = 6

// console.log(twoSum(nums, target))

// brute force, On2
const twoSum = (nums, target) => {
  let final = [];

  // looping thru array
  for (let index = 0; index < nums.length; index++) {
    // looping through the rest of the array
    for (let j = index + 1; j < nums.length; j++) {
      // if they equal target, put them in the final
      if (nums[index] + nums[j] === target) {
        final.push(index);
        final.push(j)
      }
    }
  }
  return final;
}


// better using Hash/Object
const twoSumBetter = (nums, target) => {
  let numHash = {};
  // iterate thru array, assign element as key and index as value in numHash
  for (let index = 0; index < nums.length; index++) {
    const element = nums[index];
    numHash[element] = index;
  }
  // console.log(numHash)
  // loop thru array. check if numHash has property(KEY) of diff (that is not the same as index)
  // if so, return the index and relevant other index
  for (let index = 0; index < nums.length; index++) {
    const diff = target - nums[index];
    if (numHash.hasOwnProperty(diff) && numHash[diff] !== index) {
      return [index, numHash[diff]];
    }
  }
}

let nums = [2,7,11,15]
let target = 9
console.log(twoSumBetter(nums, target))